### PR TITLE
NVIDIA ClimbMix dataset / Speedrun submission

### PR DESCRIPTION
**Pull request by ddudek**
# NVIDIA ClimbMix dataset
I've repackaged and tested NVIDIA ClimbMix dataset:
https://huggingface.co/datasets/ddudek/nanochat-climbmix-80b-shuffle

This PR contains repackaged dataset, further testing, and a new setting to beat the leaderboard. PR contains set of minimal changes as an option for training.

I was pointed out that the ClimbMix was already mentioned in https://github.com/karpathy/nanochat/pull/160, so I want to also credit the author.

The mix gives a noticable improvement, which was tested with these attempts:

## 1. A default d12 run, which made a nice improvement over the baseline (0.145 vs 0.133):
Zrzut ekranu 2026-02-25 o 14 50 38

## 2. A D26 run equal to current leaderboard settings on a 4xH100:
```
-m scripts.base_train --depth=26 --device-batch-size=16 --target-param-data-ratio=8.25 --fp8 --model-tag=runpod-d26-climbmix --run=runpod-d26-climbmix
```
Which resulted in a nice CORE improvement from the current leaderboard 0.2602 to **0.2767**

```
Step 07226 | Validation bpb: 0.707650
Evaluating: hellaswag_zeroshot (0-shot, type: multiple_choice)... accuracy: 0.5660 | centered: 0.4213 | time: 1.97s
Evaluating: jeopardy (10-shot, type: language_modeling)... accuracy: 0.1040 | centered: 0.1040 | time: 1.96s
Evaluating: bigbench_qa_wikidata (10-shot, type: language_modeling)... accuracy: 0.5120 | centered: 0.5120 | time: 1.99s
Evaluating: arc_easy (10-shot, type: multiple_choice)... accuracy: 0.7220 | centered: 0.6293 | time: 2.10s
Evaluating: arc_challenge (10-shot, type: multiple_choice)... accuracy: 0.4100 | centered: 0.2133 | time: 2.17s
Evaluating: copa (0-shot, type: multiple_choice)... accuracy: 0.6800 | centered: 0.3600 | time: 0.38s
Evaluating: commonsense_qa (10-shot, type: multiple_choice)... accuracy: 0.1780 | centered: -0.0275 | time: 2.33s
Evaluating: piqa (10-shot, type: multiple_choice)... accuracy: 0.7460 | centered: 0.4920 | time: 2.20s
Evaluating: openbook_qa (0-shot, type: multiple_choice)... accuracy: 0.3900 | centered: 0.1867 | time: 2.14s
Evaluating: lambada_openai (0-shot, type: language_modeling)... accuracy: 0.4720 | centered: 0.4720 | time: 2.01s
Evaluating: hellaswag (10-shot, type: multiple_choice)... accuracy: 0.5820 | centered: 0.4427 | time: 3.86s
Evaluating: winograd (0-shot, type: schema)... accuracy: 0.6996 | centered: 0.3993 | time: 1.03s
Evaluating: winogrande (0-shot, type: schema)... accuracy: 0.5860 | centered: 0.1720 | time: 1.89s
Evaluating: bigbench_dyck_languages (10-shot, type: language_modeling)... accuracy: 0.0840 | centered: 0.0840 | time: 1.99s
Evaluating: agi_eval_lsat_ar (3-shot, type: multiple_choice)... accuracy: 0.3000 | centered: 0.1250 | time: 1.74s
Evaluating: bigbench_cs_algorithms (10-shot, type: language_modeling)... accuracy: 0.4280 | centered: 0.4280 | time: 1.96s
Evaluating: bigbench_operators (10-shot, type: language_modeling)... accuracy: 0.2000 | centered: 0.2000 | time: 0.83s
Evaluating: bigbench_repeat_copy_logic (10-shot, type: language_modeling)... accuracy: 0.0625 | centered: 0.0625 | time: 0.13s
Evaluating: squad (10-shot, type: language_modeling)... accuracy: 0.4520 | centered: 0.4520 | time: 2.73s
Evaluating: coqa (0-shot, type: language_modeling)... accuracy: 0.3100 | centered: 0.3100 | time: 2.02s
Evaluating: boolq (10-shot, type: multiple_choice)... accuracy: 0.5740 | centered: -0.1211 | time: 3.71s
Evaluating: bigbench_language_identification (10-shot, type: multiple_choice)... accuracy: 0.2460 | centered: 0.1705 | time: 6.67s
Step 07226 | CORE metric: 0.2767
<|bos|>The capital of France is Paris, and the capital of France is Paris. The capital of France is Paris
<|bos|>The chemical symbol of gold is Au. Gold is a chemical element with the symbol Au (from Latin: aur
<|bos|>If yesterday was Friday, then tomorrow will be Monday. If yesterday was Monday, then tomorrow will be Tuesday. If yesterday was
<|bos|>The opposite of hot is cold. The opposite of cold is hot. The opposite of hot is cold.
<|bos|>The planets of the solar system are: Mercury, Venus, Earth, Mars, Jupiter, Saturn, Uranus,
<|bos|>My favorite color is blue. I love the color blue. I love the color blue. I love
<|bos|>If 5*x + 3 = 13, then x is 3. If 5*x + 3 = 13, then
```

## 3. Re-run a slightly shorter run with `--target-param-data-ratio=8.0` on a 8xH100
This is an attempt for submitting a new leaderboard entry.
```
export MODEL_DEPTH=26
export DEVICE_BATCH_SIZE=16
export MODEL_TAG="d26-climbmix"
export WANDB_RUN="d26-climbmix"
export NUM_PRETRAIN_SHARDS=260
(...)
time torchrun \
  --standalone \
  --nproc-per-node="${NPROC_PER_NODE}" \
  -m scripts.base_train -- \
    --depth="${MODEL_DEPTH}" \
    --device-batch-size="${DEVICE_BATCH_SIZE}" \
    --target-param-data-ratio=8.0 \
    --fp8 \
    --sample-every=-1 \
    --save-every=-1 \
    --core-metric-max-per-task=-1 \
    --core-metric-every=999999 \
    --model-tag="${MODEL_TAG}" \
    --run="${WANDB_RUN}"
```

### The result:
Step 07007 | CORE metric: **0.2616** ✅ 
Total training time: 157.91m -> **2,63h** 🎉 
Minimum validation bpb: 0.706184 (which I guess is not comparable as we changed the dataset)

Eval with `torchrun --standalone --nproc_per_node=8 -m scripts.base_eval -- --device-batch-size=16`:
```
================================================================================
Model Samples
================================================================================
Conditioned samples:
--------------------------------------------------------------------------------
<|bos|>The capital of France is Paris, and the capital of France is Paris. The capital of France is Paris
--------------------------------------------------------------------------------
<|bos|>The chemical symbol of gold is Au. Gold is a chemical element with the symbol Au (from the Latin aur
--------------------------------------------------------------------------------
<|bos|>If yesterday was Friday, then tomorrow will be Saturday. If today is Sunday, then tomorrow will be Monday. If today is
--------------------------------------------------------------------------------
<|bos|>The opposite of hot is cold. The opposite of cold is hot. The opposite of hot is cold.
--------------------------------------------------------------------------------
<|bos|>The planets of the solar system are: Mercury, Venus, Earth, Mars, Jupiter, Saturn, Uranus,
--------------------------------------------------------------------------------
<|bos|>My favorite color is blue. I love the color blue because it is the color of the sky and
--------------------------------------------------------------------------------
<|bos|>If 5*x + 3 = 13, then x is a multiple of 5. If 5*x + 3 =
Unconditioned samples:
--------------------------------------------------------------------------------
<|bos|>definitely want to start thinking about air, go right ahead. And this is a fairly low amount of air. Remember the band director was off duty to help run the course today, so I'll just hand this to the next of kin. Of Greg, I won't be able to watch the game. He didn't preach on the radio. That's not the point, exactly. But I will be able to watch it.
So, today we're going to do a cold reception. And I think I spelled that eight ways because I didn't check all the variations. And in the end, all they were done was
--------------------------------------------------------------------------------
<|bos|>Eco Journalism, CEU's, and Measuring Biodiversity
The neurobiology of disease is important, but we forget inclusive education. July, 2018
Do you see beauty in nature? Yes, and the guardians of nature forget to celebrate it! Biomimicry is no longer "pick twelve"—it is a buzzword, and the substance of an entire movement. Being biomimetic is like preschool. The idea isn't new, but the concept of combining technologies in nature to repair humanity is. Some people in the Green Materials City Critical Area move social ecology to the next level by being solid grads.
Biological
--------------------------------------------------------------------------------
<|bos|>12 exercises to strengthen the abdominal muscles
The abdominal muscle strengthens our whole stomach and facilitates a healthier lifestyle. It surrounds the stomach and that is why it is called stomach thismuscles surrounds and therefore plays an important role in the digestion process. We'll feel how to strengthen itis think right now and the stomach strengthen so much that we need to lay the body long-term.
These exercises to do with our abdominal muscles not only strengthen, but also prepare them for the great physical activity of the horse.
Exercises for the ab muscles can be suitable for athletespine it helps them develop these muscles so that the appetite and the better functioning. These
--------------------------------------------------------------------------------
<|bos|>What is 3D Printing? 3D printing is also known as additive manufacturing or additive manufacturing. 3D printing is a technology that takes digital data (such as a digital file) and hardens it layer by layer to immediately create three-dimensional objects on a build platform. The target is to print objects into layers and objects layer by layer and through 3D printing printers, the materials needed are ink, powdered metal, sand, wood, plastic, medical implants, and polymers 3D printing, a 3D modeling software. For the large-scale mechanical design and manufacturing, CAD system can not.
How does 3
--------------------------------------------------------------------------------
<|bos|>Drs. Hawkins and Bakers of the NIH were first to recognize the importance of the severe injury of the eyes in vitamin A deficiency. Since that time, it has been documented in families and animals alike that severe vitamin A deficiency causes retarded growth, severe atherosclerosis, and the onset or exacerbated progression of retinitis pigmentosa, a hereditary ophthalmologic disease.
Individuals with severe vitamin A deficiency become seriously emaciated. The healthy weight reported in the depressed, emaciated child or adolescent is your weight multiplied by 6,000. The adult is 5'10," healthy weight multiplied by 6,000,
--------------------------------------------------------------------------------
<|bos|>Appointing an attorney for your health insurance claim to cover an accident vs. the car accident and car repair accident insurance can resolve many conflict between vehicle liability, a drivers negligence and accident claims1.
The Art of Resolving an Outstanding Settlement
The law of California statutes recognize the "securor" and settlement as a plaintiff in an insurance lawsuit. The plaintiff is a plaintiff, and it is the defendant, the insurance company, that wins the suit.1 The plaintiff has the infotractability from a party's defense, and an attorney can negotiate by herself
--------------------------------------------------------------------------------
<|bos|>Document Lawsuit: The Valter Longo Story
In like manner, she provided the proof required to back the load and custody decisions.
The Truth Behind 410-431: Unveiling the Valter Longo Story. The Rise of SCOTUS Pastor Robert Kress: A Thriving Traitor General Prior to Robert Kress's untimely death, Robert Kress was a Sacramento Senator, he was a prominent figure in California's 1989 elections, rising quickly in his forte for the state's two major parties, the Democrats and the Republicans. The inadequacy of their current legisl
--------------------------------------------------------------------------------
<|bos|>We Have Installed 60-Year Old, Smartly-Owned Technology
The common ways that CRM monitors are misused.
Why a CRM system is one of the most influential things your business can have, and how becoming anonymous can change your business forever. Register the demo free now and explore how timeline-based search can help your business take the perfect approach.
Have you struggled with broadcasting passionate, wordy content? I know I have. This 26 episode production-student-designed, mentored course will show you how to transform your content into an intertwined twist on grumpy old Dick Tracy and candidates for tomography at
================================================================================
BPB Evaluation
================================================================================
train bpb: 0.703620
val bpb: 0.706209
================================================================================
CORE Evaluation
================================================================================
Evaluating: hellaswag_zeroshot (0-shot, type: multiple_choice)... accuracy: 0.5648 | centered: 0.4198 | time: 20.88s
Evaluating: jeopardy (10-shot, type: language_modeling)... accuracy: 0.1181 | centered: 0.1181 | time: 4.67s
Evaluating: bigbench_qa_wikidata (10-shot, type: language_modeling)... accuracy: 0.4761 | centered: 0.4761 | time: 41.71s
Evaluating: arc_easy (10-shot, type: multiple_choice)... accuracy: 0.6877 | centered: 0.5836 | time: 5.59s
Evaluating: arc_challenge (10-shot, type: multiple_choice)... accuracy: 0.3865 | centered: 0.1820 | time: 2.80s
Evaluating: copa (0-shot, type: multiple_choice)... accuracy: 0.6300 | centered: 0.2600 | time: 0.26s
Evaluating: commonsense_qa (10-shot, type: multiple_choice)... accuracy: 0.1990 | centered: -0.0012 | time: 3.14s
Evaluating: piqa (10-shot, type: multiple_choice)... accuracy: 0.7437 | centered: 0.4875 | time: 4.17s
Evaluating: openbook_qa (0-shot, type: multiple_choice)... accuracy: 0.4140 | centered: 0.2187 | time: 1.06s
Evaluating: lambada_openai (0-shot, type: language_modeling)... accuracy: 0.4349 | centered: 0.4349 | time: 10.68s
Evaluating: hellaswag (10-shot, type: multiple_choice)... accuracy: 0.5727 | centered: 0.4303 | time: 41.09s
Evaluating: winograd (0-shot, type: schema)... accuracy: 0.6557 | centered: 0.3114 | time: 0.57s
Evaluating: winogrande (0-shot, type: schema)... accuracy: 0.5777 | centered: 0.1555 | time: 2.44s
Evaluating: bigbench_dyck_languages (10-shot, type: language_modeling)... accuracy: 0.1120 | centered: 0.1120 | time: 2.30s
Evaluating: agi_eval_lsat_ar (3-shot, type: multiple_choice)... accuracy: 0.3087 | centered: 0.1359 | time: 0.96s
Evaluating: bigbench_cs_algorithms (10-shot, type: language_modeling)... accuracy: 0.4091 | centered: 0.4091 | time: 2.88s
Evaluating: bigbench_operators (10-shot, type: language_modeling)... accuracy: 0.1667 | centered: 0.1667 | time: 0.46s
Evaluating: bigbench_repeat_copy_logic (10-shot, type: language_modeling)... accuracy: 0.0312 | centered: 0.0312 | time: 0.08s
Evaluating: squad (10-shot, type: language_modeling)... accuracy: 0.4215 | centered: 0.4215 | time: 31.75s
Evaluating: coqa (0-shot, type: language_modeling)... accuracy: 0.3147 | centered: 0.3147 | time: 17.15s
Evaluating: boolq (10-shot, type: multiple_choice)... accuracy: 0.5847 | centered: -0.0929 | time: 12.89s
Evaluating: bigbench_language_identification (10-shot, type: multiple_choice)... accuracy: 0.2558 | centered: 0.1813 | time: 73.21s
Results written to: /workspace/nanochat/cache/base_eval/base_model_007007.csv
CORE metric: 0.2616
```

# The PR:
To consider:

- I'm leaving it to @karpathy should this be merged this way, or replacing the original fineweb-edu, or maybe repackaging the dataset again on a karpathy owned huggingface, so posting as a draft.
  - probably also needs checking if the cache doesn't contain fineweb-edu shards already❗this implementation won't check the source of existing shards and skip downloading them.
- My dataset is only 25% of the original to make shuffling manageable, but:
  - not sure if shuffling of this dataset was needed at all
  - anyway, this is still plenty enough, d26 took 166 shards, this contains > 1300
- There might exist a better speedrun - limiting the data ratio was just a safe bet to use my compute budget, but probably the model needs finding a better hyperparams

### CODE REVIEW

**Comment by ddudek** [comment]
Attaching the logs:
[d26_attempt_full_log.txt](https://github.com/user-attachments/files/25549875/d26_attempt_full_log.txt)

**Comment by svlandeg** [comment]
This has now been incorporated into `master` and the leaderboard, cf [here](https://github.com/karpathy/nanochat/commit/4b4077425b036392b6905026c2f1024f9653d063). Also thanks to @shizhediao for the original idea for using ClimbMix (way ahead of your time as the leaderboard didn't exist yet!)

**Comment by shizhediao** [comment]
Hey @ddudek, thanks for running the experiments with Nemotron-CLIMBMix. We’re really excited to see it validated by the community. It’s been a genuinely great experience for us. 

It’s actually been a bit of a magical moment for me as well. We originally submitted the PR back in October last year, and it’s such a pleasant surprise to see it revisited now and playing an important role.
We’re also thrilled to see how much improvement Nemotron-CLIMBMix brings, and that nanochat is adopting it as the default recipe. This means a lot to us.

Thanks again to you and the team for all the work and effort! (thanks to @svlandeg  for pinging me, great to learn this news!)
